### PR TITLE
SQL: Allow look-ahead resolution of aliases for WHERE clause

### LINKED ISSUE #29983: SQL: WHERE clause doesn't utilise alias of SELECT expression/func successfully
*Original comment by @gingerwizard:*

Query as follows:

`SELECT respondent_id, applicant_income_000s/loan_amount_000s AS income_to_earnings_ratio FROM mortgages WHERE income_to_earnings_ratio > 1 ORDER BY income_to_earnings_ratio`

The WHERE clauses here uses an alias. The query complains the income_to_earnings_ratio cannot be found i.e.

```
Found 1 problem(s)
line 1:101: Unknown column [income_to_earnings_ratio]
```

This is inconsistent with our ability to use aliases in GROUP BY and ORDER BY. Repeating the expression in the WHERE works.

### PR DESCRIPTION
**Pull request by costin**
Aliases defined in SELECT (Project or Aggregate) are now resolved in the
following WHERE clause. The Analyzer has been enhanced to identify this
rule and replace the field accordingly so that queries such as:

SELECT int AS i FROM t WHERE i > 10

Do not fail anymore (as i was unresolved inside the `WHERE` clause).

Close #29983

### CODE REVIEW

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java [hunk 1]
```diff
@@ -29,6 +30,7 @@
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.singletonMap;
 
+@TestLogging("org.elasticsearch.xpack.sql:TRACE")
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: Is this on purpose?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 2]
```diff
@@ -762,6 +763,67 @@ private static UnresolvedAttribute resolveMetadataToMessage(UnresolvedAttribute
...
+                        }
+                    }
+                }
+            }
+
+            return plan;
+        }
+
+        private Expression replaceAliases(Expression condition, List<? extends NamedExpression> named) {
+            List<Alias> aliases = new ArrayList<>();
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Wouldn't be more efficient to have a `Map<String, Alias>` where the key is the `alias.name()` and avoid iteration over a list?

**Review by matriv — APPROVED** [review]
LGTM. Nice, this typically is not supported by RDBMS (MySQL, PostgreSQL, etc.)
```
postgres=# SELECT emp_no AS a FROM test_emp WHERE a > 10;
ERROR:  column "a" does not exist
LINE 1: SELECT emp_no AS a FROM test_emp WHERE a > 10;                                        ^
````

The reason is that `SELECT` is supposed to wrap the `FROM t WHERE...` so `WHERE` is unaware of the projection. But anyway, it's handy for the user to have it.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: It is; will update.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java (hunk 1 above)
Comment (costin) [reply x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/VerifierErrorMessagesTests.java]: No, leftover.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java (hunk 2 above)
Comment (costin) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Actually scratch that - it's not that easy due to the qualifier which might be present or not. I'll update the resolution though.

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 3]
```diff
@@ -762,6 +763,67 @@ private static UnresolvedAttribute resolveMetadataToMessage(UnresolvedAttribute
...
+            named.forEach(n -> {
+                if (n instanceof Alias) {
+                    aliases.add((Alias) n);
+                }
+            });
+
+            return condition.transformDown(u -> {
+                for (Alias alias : aliases) {
+                    if (alias.name().equals(u.name()) || Objects.equals(alias.qualifiedName(), u.qualifiedName())) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Is there a case where the names are not equal but the `qualifiedName()` are equal?
Couldn't we check just the `qualifiedName()`?
If there is such case can we have it in a test?

**Comment by costin** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by costin** [comment]
@gingerwizard FYI

**Pull request merged**